Add print_rich() for printing with BBCode

voylin opened the pull request:
This is an attempt to make pull request #33541 work.

The addition is a 'print_rich()' function which allows the use of BBCode for printing. It prints in both in color, bold, italic, underline in the editor output and in the console through ANSI. (With the help of @Calinou for the conversion strings)

Like you can see in this screenshot, when using print(...) everything gets printed like you've typed it.
But when you use print_rich(...), BBCodes will be used.
![Screenshot from 2022-05-02 13-50-53](https://user-images.githubusercontent.com/56970759/166186982-198cf95d-48f0-42b4-80b4-0cfe0b3446b5.png)

I was also able to make the folding work that happens with the other types of logs aswell.


Sorry in advance if the way I made this work isn't very clean, I hope you guys can give me advice on how to improve this code. I tried my best and this is my first time actually working on the Godot engine. But it works. 😄

voylin commented:
I think I fixed all the workflow issues (ran the scripts on my pc and it worked), and re-based on the master branch.

kleonc commented on core/string/print_string.cpp:
```diff
@@ -74,6 +74,52 @@ void __print_line(String p_string) {
+	p_string = p_string.replace("[/u]", "");
+	p_string = p_string.replace("[s]", "");
+	p_string = p_string.replace("[/s]", "");
+	p_string = p_string.replace("[indent]", "");
+	p_string = p_string.replace("[/indent]", "");
+	p_string = p_string.replace("[code]", "");
+	p_string = p_string.replace("[/code]", "");
+	p_string = p_string.replace("[center]", "");
+	p_string = p_string.replace("[/center]", "");
+	p_string = p_string.replace("[right]", "");
+	p_string = p_string.replace("[/right]", "");
+	p_string = p_string.replace("[color=black]", "");
+	p_string = p_string.replace("[color=red]", "");
+	p_string = p_string.replace("[color=green]", "");
+	p_string = p_string.replace("[color=lime]", "");
+	p_string = p_string.replace("[color=yellow]", "");
+	p_string = p_string.replace("[color=blue]", "");
+	p_string = p_string.replace("[color=magenta]", "");
+	p_string = p_string.replace("[color=pink]", "");
+	p_string = p_string.replace("[color=purple]", "");
+	p_string = p_string.replace("[color=cyan]", "");
+	p_string = p_string.replace("[color=white]", "");
+	p_string = p_string.replace("[color=orange]", "");
+	p_string = p_string.replace("[color=gray]", "");
+	p_string = p_string.replace("[/color]", "");
+	p_string = p_string.replace("[bgcolor=black]", "");
+	p_string = p_string.replace("[bgcolor=red]", "");
+	p_string = p_string.replace("[bgcolor=green]", "");
+	p_string = p_string.replace("[bgcolor=lime]", "");
+	p_string = p_string.replace("[bgcolor=yellow]", "");
+	p_string = p_string.replace("[bgcolor=blue]", "");
+	p_string = p_string.replace("[bgcolor=magenta]", "");
+	p_string = p_string.replace("[bgcolor=pink]", "");
+	p_string = p_string.replace("[bgcolor=purple]", "");
+	p_string = p_string.replace("[bgcolor=cyan]", "");
+	p_string = p_string.replace("[bgcolor=white]", "");
+	p_string = p_string.replace("[bgcolor=orange]", "");
+	p_string = p_string.replace("[bgcolor=gray]", "");
+	p_string = p_string.replace("[/bgcolor]", "");
+
```
Normal printing shouldn't care about these. If I'd write `print("[b]")` I'd expect `[b]` to be printed etc.

```suggestion
```

voylin replied:
The problem with this is that else you can also use bbcode in normal print. How could we solve that problem then?

voylin replied:
@kleonc How about the update which I've done, it prints [b] and such in the terminal but still removes it for the editor console. I can't really find another way of doing this, else the normal print("[b]") also makes the text bold because the text gets appended. I tried figuring out how to make the text only append for print_rich(), but they all failed as I could not give them a new p_type like MSG_TYPE_RICH. All messages to the log get MSG_TYPE_STD no matter what I did.

So I hope this is okay that it prints the raw bbcode in the terminal but not in the editor console. I updated the picture in the first comment with an example of how it looks now.

kleonc replied:
Oh, you mean that `print("[b]text[/b]")` would print bold `text` in the editor because in-editor console is a `RichTextLabel` with bbcode enabled? Hmm, that's unfortunate. I don't think `RichTextLabel` allows to ignore bbcode in some part of its assigned text (it either parses all or nothing?:thinking:). So to solve it properly probably `RichTextLabel` would need to be modified to somehow allow partially disabling bbcode parsing. Not sure. Alternatively multiple `RichTextLabel`s could be used (so only some of them would parse bbcode) but I don't think it would work well, it would probably cause problems with text selecting etc.
So I don't see a simple solution.

But again: if `print("[b]text[/b]")` will print just `text` then I think it's a problem.

kleonc replied:
> @kleonc How about the update which I've done, it prints [b] and such in the terminal but still removes it for the editor console. I can't really find another way of doing this, else the normal print("[b]") also makes the text bold because the text gets appended. I tried figuring out how to make the text only append for print_rich(), but they all failed as I could not give them a new p_type like MSG_TYPE_RICH. All messages to the log get MSG_TYPE_STD no matter what I did.
> 
> So I hope this is okay that it prints the raw bbcode in the terminal but not in the editor console. I updated the picture in the first comment with an example of how it looks now.

Hmm, I think it might be acceptable/good enough if this will be well documented. Solving it properly seems not so straighforward and I'm not sure if it's really worth the effort.
Of course that's just my opinion, I'm not anyhow decisive here.

voylin replied:
Well I kind of find that it should be like how you say, when doing print("[b]") that it actually print [b].
It's not easy but I'm slowly getting to a solution on how to solve this. However the new console system in the editor is putting up a fight as I can't just create a new type of message. Not sure yet how to solve it, but I'm halfway there. ^^

voylin replied:
@kleonc I got good news! I updated the first comment if you want to see more of the information. But basically I was able to make it so print(...) prints everything in a normal way again like it does now. And only when using print_rich(...) you will be able to use BBCodes.

I hope you are happy with this change. I am most certainly happy as I can now say that I did a good job. This is my first contribution so I hope it gets accepted, but I'm already quite happy with what I achieved. Thanks for pointing this issue out, it motivated me to try one more time to improve this whole pull request. Thanks! 😄

kleonc replied:
Nice, thanks! Seems like logging is way more customizable than I thought (yeah, I'm not familiar with the relevant code :laughing:).

Yes, I'm happy with this change behavior-wise. Can't really comment on the conceptual side of the code (as I'm not familiar with it and I didn't look into it deeply).

voylin commented:
I squashed all my commits, re-based them on master, did a check with clang_format.sh and also fixed an issue with one of the Linux workflow builds (made a mistake in the mono stuff, the function name wasn't correct). So normally all workflow checks should be good now.

voylin commented:
Fixed the workflow issue which was happening. Sorry about that. ^^"

yuyutsupoudel commented:
This is really nice. Thanks for implementing it. Hope added soon.

kleonc commented on editor/editor_log.h:
```diff
@@ -48,6 +48,7 @@ class EditorLog : public HBoxContainer {
 	enum MessageType {
 		MSG_TYPE_STD,
 		MSG_TYPE_ERROR,
+		MSG_TYPE_RICH,
 		MSG_TYPE_WARNING,
 		MSG_TYPE_EDITOR,
 	};
```
Same as above: isn't `MSG_TYPE_RICH` just a "subtype" of `MSG_TYPE_STD`? If that's the case then maybe naming it `MSG_TYPE_STD_RICH` would make more sense?

voylin replied:
Made the changes and pushed out the rebase. Now hoping that the workflows will still all be positive hahaha, seeing those run is starting to give me stress hahaha

kleonc replied:
Based on [your other comment](https://github.com/godotengine/godot/pull/60675#issuecomment-1114522170) it seems like you're trigerring format checks manually. You can [add pre-commit hooks](https://docs.godotengine.org/en/3.4/community/contributing/code_style_guidelines.html#pre-commit-hook) so such checks will be automatically made locally  for you whenever you commit. This way you won't be afraid of failing static checks (not sure what else could fail after such simple changes :wink:).

voylin replied:
@kleonc I didn't know about those pre-commit hooks, I always just used the clang_format.sh, I will take this in mind next time. I have been learning a lot about git and github recently. 
I've got one more question, now the workflow checks are completed, how does the process go from here? Just wait until someone approves the merge or gives comments if things needs to change?

kleonc replied:
>I've got one more question, now the workflow checks are completed, how does the process go from here? Just wait until someone approves the merge or gives comments if things needs to change?

You could just wait or you could ask for a review on the [Godot Contributors Chat](https://chat.godotengine.org) (probably in the `general` channel). If some core contributor(s) would approve and no one would have anything against then who knows maybe it could be merged even today.

voylin commented on core/debugger/remote_debugger.h:
```diff
@@ -44,6 +44,7 @@ class RemoteDebugger : public EngineDebugger {
 	enum MessageType {
 		MESSAGE_TYPE_LOG,
 		MESSAGE_TYPE_ERROR,
+		MESSAGE_TYPE_RICH,
 	};
```
I think I better change it to make sure it is understandable for others when they need to edit this in the future, I will make the changes, rebase and push.

Merlin1846 commented:
I don't think this would work with all terminals, which makes this not usable in the final export, right?

voylin commented:
> I don't think this would work with all terminals, which makes this not usable in the final export, right?

It uses ANSI so it should be supported by nearly all terminals I think, but I could be wrong of course. For which terminals would this not work?

Calinou commented:
> I don't think this would work with all terminals, which makes this not usable in the final export, right?

- Support on Linux and macOS is pretty good.
- Support on Windows 10 and later requires https://github.com/godotengine/godot/pull/44118 to be merged.
- Windows 8 and prior will not support this. There should be a routine to replace BBCode with empty strings, which should be used when Windows 8 or prior is detected. This doesn't have to be done now; this can be implemented in a future pull request.

miv391 commented:
There are 44 `p_string_ansi.replace` calls so I started thinking whether all the bbcode replaces could be done with one pass. I don't how much faster one pass implementation could be and I don't know if it even matters for the intended usage and would it be worth of much more complicated code.

Calinou commented:
> There are 44 `p_string_ansi.replace` calls so I started thinking whether all the bbcode replaces could be done with one pass. I don't how much faster one pass implementation could be and I don't know if it even matters for the intended usage and would it be worth of much more complicated code.

I'd say the set of `replace()` calls is fast enough right now (see my screenshot in https://github.com/godotengine/godot/pull/33541#issuecomment-1114006020).

If making it faster is desired, we can check for the presence of a tag before running its relevant replacements. For instance, if there is no `[fgcolor` within the string, there's no point in running all the `.replace()` calls for `[fgcolor` tags. The same applies to `[color` and `[bgcolor`.

voylin commented:
> There are 44 `p_string_ansi.replace` calls so I started thinking whether all the bbcode replaces could be done with one pass. I don't how much faster one pass implementation could be and I don't know if it even matters for the intended usage and would it be worth of much more complicated code.

Like Calinou suggested, I've added the 2 biggest pieces ([color=*] and [bgcolor=*]) inside of an if statement to see if the string contains any color or bgcolor tags. Hope this helps with the speed concern.

voylin committed:
Adding print_rich for printing with BBCode

raulsntos commented on modules/mono/glue/GodotSharp/GodotSharp/Core/GD.cs:
```diff
@@ -238,6 +238,14 @@ public static void Print(params object[] what)
+        /// <summary>
+        /// Prints a string including BBCode
+        /// </summary>
+        public static void Print_rich(params object[] what)
```
The C# convention is to use PascalCase for method names.

```suggestion
        public static void PrintRich(params object[] what)
```

voylin replied:
Fixed this in my latest comment, thanks for letting me know!

raulsntos commented on doc/classes/@GlobalScope.xml:
```diff
@@ -646,6 +646,16 @@
+		<method name="print_rich" qualifiers="vararg">
+			<description>
+				Converts one or more arguments of any type to string in the best way possible and prints them to the console. Using BBCode for colors, italics and bold is possible.
+				[codeblock]
+				var a = [1, 2, 3]
+				print("a", "b", a) # Prints ab[1, 2, 3]
+				[/codeblock]
```
This example uses `print` instead of `print_rich`. I would also add some BBCode to the arguments.

voylin replied:
Made a fitting example which uses some BBCode instead and also changed it to print_rich.

raulsntos replied:
You forgot the quotes for the string argument in the example.

raulsntos commented on modules/mono/glue/GodotSharp/GodotSharp/Core/GD.cs:
```diff
@@ -238,6 +238,14 @@ public static void Print(params object[] what)
             godot_icall_GD_print(GetPrintParams(what));
         }
 
+        /// <summary>
+        /// Prints a string including BBCode
+        /// </summary>
```
We try to keep the C# documentation in sync with the GDScript documentation:

```suggestion
        /// <summary>
        /// Converts one or more arguments of any type to string in the best way possible
        /// and prints them to the console. Using BBCode for colors, italics and bold is possible.
        ///
        /// Note: Consider using <see cref="PushError(string)"/> and <see cref="PushWarning(string)"/>
        /// to print error and warning messages instead of <see cref="Print(object[])"/>.
        /// This distinguishes them from print messages used for debugging purposes,
        /// while also displaying a stack trace when an error or warning is printed.
        /// </summary>
        /// <example>
        /// <code>
        /// var a = new int[] { 1, 2, 3 };
        /// GD.PrintRich("a", "b", a); // Prints ab[1, 2, 3]
        /// </code>
        /// </example>
        /// <param name="what">Arguments that will be printed.</param>
```

I noticed that the documentation you added for GDScript uses `print` instead of `print_rich` in the example, so if you change it make sure to update it here too.

voylin replied:
Thank you for this! Wasn't really sure how to do the documentation for the c# part. I changed the example to the same one being used in @GlobalScope.

raulsntos replied:
You still left code from the previous example, and forgot the quotes for the string argument.

voylin replied:
I'm so sorry about that, I fixed it and added the quotes both here and in the GlobalScope.
Thank you for the help and sorry again for missing this!

raulsntos replied:
It's fine, don't worry about it.

voylin commented:
The static checks failed because of the color tags I used in the example for the docs, so I changed the example. There was also a small mistake in the GlobalScope documentation which I have changed.

Calinou commented on editor/editor_node.cpp:
```diff
@@ -5802,9 +5802,10 @@ static Node *_resource_get_edited_scene() {
+void EditorNode::_print_handler(void *p_this, const String &p_string, bool p_error, bool p_rich) {
 	EditorNode *en = static_cast<EditorNode *>(p_this);
-	en->log->add_message(p_string, p_error ? EditorLog::MSG_TYPE_ERROR : EditorLog::MSG_TYPE_STD);
+	en->log->add_message(p_string, p_error ? EditorLog::MSG_TYPE_ERROR : p_rich ? EditorLog::MSG_TYPE_STD_RICH
+																				: EditorLog::MSG_TYPE_STD);
```
Avoid nested ternary operator as described above:

```suggestion
	if (p_error) {
		en->log->add_message(p_string, EditorLog::MSG_TYPE_ERROR);
	} else if (p_rich) {
		en->log->add_message(p_string, EditorLog::MSG_TYPE_STD_RICH);
	} else {
		en->log->add_message(p_string, EditorLog::MSG_TYPE_STD);
	}
```

Calinou commented on core/debugger/remote_debugger.cpp:
```diff
@@ -237,7 +237,8 @@ void RemoteDebugger::_print_handler(void *p_this, const String &p_string, bool p
 		OutputString output_string;
 		output_string.message = s;
-		output_string.type = p_error ? MESSAGE_TYPE_ERROR : MESSAGE_TYPE_LOG;
+		output_string.type = p_error ? MESSAGE_TYPE_ERROR : p_rich ? MESSAGE_TYPE_LOG_RICH
+																   : MESSAGE_TYPE_LOG;
```
Avoid nesting ternary operators, as it's not very readable:

```suggestion
		if (p_error) {
			output_string.type = MESSAGE_TYPE_ERROR;
		} else if (p_rich) {
			output_string.type = MESSAGE_TYPE_LOG_RICH;
		} else {
			output_string.type = MESSAGE_TYPE_LOG;
		}
```

voylin replied:
Sorry, I thought less lines is better. ^^"

voylin commented on core/string/print_string.cpp:
```diff
@@ -79,7 +79,80 @@ void __print_line(String p_string) {
+		p_string_ansi = p_string_ansi.replace("[/bgcolor]", "\u001b[49m");
+	}
+
+	OS::get_singleton()->print_rich("%s\n", p_string_ansi.utf8().get_data()); // Over here I need to change p_string stuff
```
Oops completely forgot that was there. ^^"

voylin commented on core/string/print_string.cpp:
```diff
@@ -79,7 +79,80 @@ void __print_line(String p_string) {
+		p_string_ansi = p_string_ansi.replace("[bgcolor=orange]", "\u001b[48;5;208m");
+		p_string_ansi = p_string_ansi.replace("[bgcolor=gray]", "\u001b[100m");
+		p_string_ansi = p_string_ansi.replace("[/bgcolor]", "\u001b[49m");
+	}
```
Added in the most recent commit!
Thank you for the help with the ANSI codes.

voylin commented on doc/classes/@GlobalScope.xml:
```diff
@@ -646,6 +646,15 @@
 		</method>
+		<method name="print_rich" qualifiers="vararg">
+			<description>
+				Converts one or more arguments of any type to string in the best way possible and prints them to the console. Using BBCode for colors, italics and bold is possible.
```
I've added this to global scope and also the the mono documentation file.

Calinou commented on editor/editor_log.cpp:
```diff
@@ -419,6 +428,11 @@ EditorLog::EditorLog() {
 	vb_right->add_child(std_filter->toggle_button);
 	type_filter_map.insert(MSG_TYPE_STD, std_filter);
 
+	LogFilter *rich_filter = memnew(LogFilter(MSG_TYPE_STD_RICH));
+	rich_filter->initialize_button(TTR("Toggle visibility of rich output messages."), callable_mp(this, &EditorLog::_set_filter_active));
+	vb_right->add_child(rich_filter->toggle_button);
+	type_filter_map.insert(MSG_TYPE_STD_RICH, rich_filter);
```
Should rich logs be considered a different type of logging for the purpose of filtering? I'm not sure if users expect it to be considered a different message category.

Instead, I'd make the standard log button toggle the display of both non-rich and rich prints.

voylin replied:
Okay, I work on it as soon as I can! :)

voylin replied:
Fixed it and working perfectly like before.

Calinou reviewed (commented):
Should be good to merge (preferably after https://github.com/godotengine/godot/pull/44118), although I have a concern about the editor UX related to filtering messages – see above.

voylin commented:
> Should be good to merge (preferably after #44118), although I have a concern about the editor UX related to filtering messages – see above.

Fixed the filtering message thing. The normal print messages and print_rich messages are being filtered by the same button now.

akien-mga reviewed (approved):
Approved in PR review meeting, the code seems good and the feature is nice to have.
Would be worth a rebase to make sure it still builds and works fine before merging.

voylin commented:
@akien-mga Rebased and checked, still working. Thanks for approving this! ^^/

akien-mga commented:
Thanks! And congrats for your first merged Godot contribution :tada:

Luke-devo commented:
Best feature ever!

- [X] Done

ggppjj commented:
Hello!

I believe nested color tags work as I would expect, but they don't seem to filter from the text logs correctly. `[color=green][color=silver]*[/color].[/color]` renders to a debug console as I would expect, with the `*` being silver and the `.` being green, but the `[color=silver]` tags are being written to the godot.log, where the top-level tags aren't.

Calinou commented:
>  but the [color=silver] tags are being written to the godot.log, where the top-level tags aren't.

Nested color tags work in the editor Output panel, but there are two limitations

- Terminal printing does not handle nested color tags.


Also, terminal printing does not handle color tags other than a few basic codes. See `print_rich()`'s description in the class reference. This would need https://github.com/godotengine/godot/issues/87899 to be implemented to be resolved.

ggppjj commented:
> Nested color tags work in the editor Output panel (it's displayed using RichTextLabel), but terminal printing does not handle nested color tags as it uses a separate BBCode implementation.

To clarify, I mean on-disk plaintext file log output generated by debug builds, if that isn't the same thing in this context as terminal printing. 

Understood about the color support, though. Thanks for the tip there!

Calinou commented:
> To clarify, I mean on-disk plaintext file log output generated by debug builds, if that isn't the same thing in this context as terminal printing.

That's a consequence of the terminal printing. What you see in the log file is directly based on stdout with [ANSI escape codes stripped](https://github.com/godotengine/godot/pull/90900). Since terminal printing never replaced the `[color=silver]` tag in the first place, it's written as-is to the log file.

ggppjj commented:
Understood, thanks. I figured that's what the answer was once you clarified that silver is only accidentally working. I'll stick to supported codes for the time being.

Thanks!
